t/harness

```diff
@@ -90,6 +72,150 @@ if ($ENV{HARNESS_OPTIONS}) {
     }
 }
 
+sub _extract_tests;
+sub _extract_tests {
+    # This can probably be done more tersely with a map, but I doubt that it
+    # would be as clear
+    my @results;
+    foreach (@_) {
+        my $ref = ref $_;
+        if ($ref) {
+            if ($ref eq 'ARRAY') {
+                push @results, _extract_tests @$_;
+            } elsif ($ref eq 'HASH') {
+                push @results, _extract_tests values %$_;
+            } else {
+                die "Unknown reference type $ref";
+            }
+        } else {
+            push @results, glob $_;
+        }
+    }
+    @results;
+}
+
+my %total_time;
+
+sub _compute_tests_and_ordering($) {
+    my @tests = $_[0]->@*;
+
+    my %dir;
+    my %serials;
+    my %all_dirs;
+    my %map_file_to_dir;
+
+    if ($jobs) {
+        require App::Prove::State;
+        $state = App::Prove::State->new({ store => 'test_state' });
+        $state->apply_switch('slow', 'save');
+        # For some reason get_tests returns *all* the tests previously run,
+        # (in the right order), not simply the selection in @tests
+        # (in the right order). Not sure if this is a bug or a feature.
+        # Whatever, *we* are only interested in the ones that are in @tests
+        my %seen;
+        @seen{@tests} = ();
+        @tests = grep {exists $seen{$_} } $state->get_tests(0, @tests);
+    }
+
+    my %times;
+    if ($state) {
+        # Where known, collate the elapsed times by test name
+        foreach ($state->results->tests()) {
+            $times{$_->name} = $_->elapsed();
+        }
+    }
+
+    # Preprocess the list of tests
+    for my $file (@tests) {
+        if ($^O eq 'MSWin32') {
+            $file =~ s,\\,/,g; # canonicalize path
+        };
+
+        # Keep a list of the distinct directory names, and another list of
+        # those which contain a file whose name begins with a 0
+        if ($file =~ m! \A ( (?: \.\. / )?
+                                .*?
+                            )             # $1 is the directory path name
+                            /
+                            ( [^/]* \. (?: t | pl ) ) # $2 is the test name
+                        \z !x)
+        {
+            my $path = $1;
+            my $name = $2;
+
+            $all_dirs{$path} = 1;
+            $map_file_to_dir{$file} = $path;
+
+            # We assume that the reason a test file's name begins with a 0
+            # is to order its execution among the tests in its directory.
+            # Hence, a directory containing such files should be tested in
+            # serial order, with some exceptions hard-coded in.
+            if ($name =~ / \A 0 /x) {
+                $serials{$path} = 1;
+            }
+            elsif (defined $must_be_executed_serially{$path}) {
+                $serials{$path} = 1;
+                delete $must_be_executed_serially{$path};
+            }
+        }
+    }
+
+    # We assume that the reason a test file's name begins with a 0 is to
```

Comment:
shouldnt this be "begins with a digit"? I think this heuristic is too picky or the comment is wrong :-)

Reply:
The paradigm is that if a directory has a test file whose name begins with a 0, to execute in numeric order all the test files in that directory whose names begin with a digit.  But at least one file has to have a leading 0.  This PR doesn't change the way this has worked for quite a while now.

Reply:
Since this PR doesn't change it, then  changing it to be less fragile can wait for another PR.